onnxruntime/core/providers/rocm/tunable/gemm_rocblas.h

```diff
@@ -32,6 +33,181 @@ class RocblasHandleStreamGuard {
   hipStream_t original_stream_;
 };
 
+#if defined USE_ROCBLAS_EXTENSION_API
+
+template <typename T>
+constexpr rocblas_datatype RocBlasDataTypeFor(const T*) {
+  static_assert(sizeof(T) == -1, "Unsupported type for rocBLAS operation.");
+  // The code below should be unreachable due to the static_assert above.
+  // But the compiler doesn't like not having a return statement, so we
+  // return something sensible.
+  return rocblas_datatype_f32_r;
+}
+
+template <>
+constexpr rocblas_datatype RocBlasDataTypeFor<float>(const float*) {
+  return rocblas_datatype_f32_r;
+}
+
+template <>
+constexpr rocblas_datatype RocBlasDataTypeFor<half>(const half*) {
+  return rocblas_datatype_f16_r;
+}
+
+template <>
+constexpr rocblas_datatype RocBlasDataTypeFor<double>(const double*) {
+  return rocblas_datatype_f64_r;
+}
+
+template <>
+constexpr rocblas_datatype RocBlasDataTypeFor<BFloat16>(const BFloat16*) {
+  return rocblas_datatype_bf16_r;
+}
+
+template <typename T>
+constexpr rocblas_datatype RocBlasComputeTypeFor(const T*) {
+  static_assert(sizeof(T) == -1, "Unsupported type for rocBLAS operation.");
+  // The code below should be unreachable due to the static_assert above.
+  // But the compiler doesn't like not having a return statement, so we
+  // return something sensible.
+  return rocblas_datatype_f32_r;
+}
+
+template <>
+constexpr rocblas_datatype RocBlasComputeTypeFor<float>(const float*) {
+  return rocblas_datatype_f32_r;
+}
+
+template <>
+constexpr rocblas_datatype RocBlasComputeTypeFor<half>(const half*) {
+  return rocblas_datatype_f32_r;
```

Comment:
Why is `half` converts to `...f32...`. If it is intended, please comment about the special here.

Reply:
This is the compute type specification for ROCBlas, not the data type specification. I've observed that if I use FP16 datatype, and FP16 compute, it is _slower_ than using FP16 datatype and FP32 compute. Also, see `fpgeneric.h`: the compute type is FP32 for FP16 and BF16 datatypes. Added a comment to clarify.